Remove form handling in WP resource

oliverguenther opened the pull request:
Refactor the WP resource usage to avoid any modifications to it using the form resource. 

- [x] A `WorkPackageResource` is now always the last known state from the API and can be updated regardless of any active editing.
- [x] Removes all means to get/update the form in the work package resource
- [x] Removes the need to override the work package schema if the form is loaded

**TODOS**

- [X] Adapt timeline to use a changeset to update dates when moving in the timeline.
- [x] Fix tests  (In progress)
- [X] Test attachments

Fixes
https://community.openproject.com/wp/25972
https://community.openproject.com/wp/25495
https://community.openproject.com/wp/26023

https://community.openproject.com/wp/26019 in 539a0a9eb1afa327ccddcb59d3ec984512dbe60d

oliverguenther committed:
Create separate changeset object to provide resource and schema

oliverguenther committed:
Use changeset in timeline

oliverguenther committed:
Fix comment preview

oliverguenther committed:
Keep erroneous fields open and instead only set their errors

oliverguenther committed:
Fix copying

oliverguenther committed:
Update fields when changeset resource is updated

oliverguenther committed:
Fix setting parent in new child feature

oliverguenther committed:
Fix overridden setter/getter in multivalue select field

oliverguenther committed:
Update the form on special fields for global create

oliverguenther committed:
Fix conflict notification

oliverguenther committed:
No longer use workPackage.isMilestone while editing

oliverguenther committed:
Don't gulp errors in PromiseLike

oliverguenther committed:
Remove the overridden schema for new work packages once saved

We still have components that use a new work package and its form schema
set by the `overriddenSchema` property. We should get rid of this
and instead let them use the edit resource in the long run.

oliverguenther committed:
Allow traversal from one single-view to the other (e.g. split -> full)

This requires us to keep changes open if we're starting to edit other
fields.

oliverguenther committed:
Correct focussing on new fields when switching

oliverguenther commented:
Only one invalid-query spec is failing.

oliverguenther committed:
Let timeline maintain a single changeset in renderInfo

ulferts commented on frontend/app/components/wp-edit-form/work-package-changeset.ts:
```diff
@@ -0,0 +1,286 @@
+    // But store the existing form in case of an error.
+    // Because if we get an error, the object returned is not a form
+    // and thus lacks the links the implementation depends upon.
+    const oldForm = this.wpForm.value;
```
If I get this right, we do no longer alter the work package. As such, we should no longer have to rely on links in the form. Do we still need to juggle the old/new form?

oliverguenther replied:
It's reset only to avoid loading it again on the next try. Not sure if there is a case where this does make sense, though.

ulferts commented on frontend/app/components/wp-edit/field-types/wp-edit-boolean-field.directive.html:
```diff
@@ -1,10 +1,10 @@
        ng-focus="vm.handleUserFocus()"
        ng-blur="vm.handleUserBlur()"
-       ng-disabled="vm.workPackage.inFlight"
+       ng-disabled="vm.field.inFlight"
```
do we still need those?

oliverguenther replied:
Yes, it's a visual feature of the _currently active_ field to disable editing it. We should keep that, in order to avoid changing the value currently being processed (e.g., in the path up until PATCHing the resource)

ulferts reviewed (changes requested):
### General

I didn't spot any errors visible to the user (except the first listed below and that one is an edge case). There are technical flaws that I would like to be addressed before merging.

### Erroneous column not added
If
* one moves a work package having an erroneous property (e.g. missing cf) in the timeline
* disregards the error
* tries to edit the same work package in the table (e.g. due date)
=> erroneous column is not added.
It is added when the work package is edited in the table right away. I didn't check if the error already existed.

### Multiple form requests
When clicking the "+" icon to add a new work package in the table, three form requests are triggered. The same, interlaced by other requests happens when using the "Create" button. This is not visible to the user but increases server load and (maybe) wait time for the user.

Pressing create button:
![image](https://user-images.githubusercontent.com/617519/28779663-4ae1abd8-7604-11e7-8cf3-f944f6d0284c.png)


### Unused validation errors
Before the work package is updated/created, a form request is send. This will already include the errors the subsequent post/put will cause. In fact, if the work package is already invalid (e.g. required custom field not set), we could show the necessary fields before sending another request. We only have to ensure that the errors are displayed in a nice way so that we do not yell at the user for an error he did not cause (e.g. required cf)

This behaviour existed before already but it strikes me as even weirder now.

If we do not want to address this we might wan to get rid of the then almost (unless type change) unnecessary extra request before the actual save.

### Inconsistent changeset usage
While only the altered fields are send in the form request,
![image](https://user-images.githubusercontent.com/617519/28779623-22f38952-7604-11e7-9d6d-153695f9d5f1.png)
ever attribute the work package possesses is send on the update/create request. 
![image](https://user-images.githubusercontent.com/617519/28779634-2ed3ef14-7604-11e7-8037-db1be35f419e.png)
This seems very inconsistent with the idea of a changeset and is prone to errors as e.g. a user might override changes of another user without wanting to (e.g. when moving the wp in the timeline).

oliverguenther commented:
> Erroneous column not added

This is identical to the previous behavior. I don't think we should add more fields to the table in this case, but rather open the details view. I find it weird to request the user to edit more attributes in the table if the only thing he wants to do is move the dates.

oliverguenther commented:
> Multiple form requests

I did stumble over the wpCreate service that requests a form without it being used anywhere other than initializing the types. This should be the same behavior as in dev, but I'll look into whether we can improve that.

oliverguenther commented:
> Unused validation errors

Good point, I thought about removing the current behavior, and there probably is no better time than now.

oliverguenther committed:
Allow initialization with a new form resource

Avoids loading the form again when the single view initializes the form
editing.

oliverguenther committed:
Only add to changes once changed

oliverguenther committed:
Create resource that contains the changeset and form

oliverguenther committed:
Store only the changeset in a state, not the form itself

With the changeset being the authoritive source of the changes to a work
package, we can pass it around.

Also, a temporary resource for viewing the changes made to the work
packages is available through the editing service.

This is required since we need to show the WP resource as it were if the
changes were applied.

oliverguenther committed:
Add spec for edge case that was missed

oliverguenther committed:
Reject submission earlier when validationErrors in form

oliverguenther committed:
Fix usages of form and usage of raw values

oliverguenther committed:
Reject the form update ONLY when trying to save

oliverguenther committed:
Don't access the changeset in the wiki subclass comment field

oliverguenther committed:
Fix clearing the edit form upon cancelling the edition

oliverguenther committed:
Fix broken copy spec due to double initialization

oliverguenther committed:
Fix moving in timeline

Fixes #26019

oliverguenther commented:
@ulferts I believe I've addressed all issues and implemented all changes you requested. The submission now feels significantly faster.

oliverguenther committed:
Don't call disable just for resetting the placeholder TL element

oliverguenther commented:
Since this PR no longer redraws the timeline upon moving the WP, this also appears to fix https://community.openproject.com/projects/openproject/work_packages/26023

ulferts commented on frontend/app/components/angular/angular-injector-bridge.functions.ts:
```diff
@@ -17,6 +17,7 @@
    * factory.
    *
    * @param injectable The target to inject into
+ * @deprecated Use $injectFields instead
```
The indentation is incorrect

ulferts commented on frontend/app/components/api/api-v3/hal-resources/type-resource.service.ts:
```diff
@@ -50,6 +50,10 @@ export class TypeResource extends HalResource {
 
+  public idFromLink():string {
+    return this.$href!.split('/').pop()!;
+  }
```
we really need to extract this into shared code

ulferts commented on frontend/app/components/wp-edit-form/work-package-changeset.ts:
```diff
@@ -0,0 +1,322 @@
+  }
+
+  public startEditing(key:string) {
+  }
```
???

ulferts commented on frontend/app/components/wp-edit-form/work-package-changeset.ts:
```diff
@@ -0,0 +1,322 @@
+    let payload:any = this.workPackage.$plain();
+
+    if (hasForm) {
+      _.defaultsDeep(payload, this.wpForm.value!.payload.$source);
```
Didn't know this one. Learned something today, thanks.

ulferts reviewed (changes requested):
### General

https://community.openproject.com/projects/telekom/work_packages/25972 and maybe https://community.openproject.com/projects/telekom/work_packages/25495 (never could reproduce this) are fixed in a way I thought we wanted to avoid. The fields are now disabled while the saving is in flight. I added a debugger statement to the wp patch method. As long as I did not continue the execution, I could still activite files, but they where all disabled:

![image](https://user-images.githubusercontent.com/617519/28911634-744142ac-7831-11e7-8920-6d0ad9296bcd.png)

Seems to fix https://community.openproject.com/projects/openproject/work_packages/26023 - dragging a bar felt fine

It alters the way we handle having a work package in edit mode in the split view and in the wp table simultaneously. While we used to close both edit forms when saving one, now only the saved one is closed. I do not find the new behaviour bad however and was pleasantly surprised to not run into version conflicts. 

### Click on wp in timeline (without changes) triggers saving

With the timeline opened, clicking on a work package (just click, no dragging), triggers saving of the work package.

### Wrong dates on error

https://community.openproject.com/projects/openproject/work_packages/26019 is not fixed. We can of course fix it separately.

### Inline create retains values 
When creating a new work package inline (wp table), the values I entered into the fields (e.g. subject) are retained after the successful save.

### Changing type or project issues two form requests
On type or project changes (create or update), two form requests are sent.

oliverguenther committed:
Remove unused startEditing

We no longer need to set the current value, since we fall back to the
resource value.

oliverguenther committed:
Avoid saving the WP if changeset is empty

oliverguenther committed:
Update labels AFTER changeset reset

oliverguenther committed:
Reset form when creating a new inline row

oliverguenther committed:
Only update the form for type and project for NEW work packages

oliverguenther committed:
Extract idFromLink into HalResource

ulferts commented:
> Changing type or project issues two form requests

Fixed

ulferts commented:
> Click on wp in timeline (without changes) triggers saving

Fixed

ulferts commented:
> Wrong dates on error

Fixed

ulferts commented:
> Inline create retains values

Fixed
